packages/openid4vc-holder/src/presentations/PresentationExchangeService.ts

```diff
@@ -0,0 +1,360 @@
+            path: prefix,
+            id: descriptor.id,
+            path_nested: {
+              ...descriptor,
+              path: descriptor.path.replace('$.', `${prefix}.vp.`),
+              format: 'jwt_vc_json',
+            },
+          }
+        })
+      )
+    }
+
+    return {
+      verifiablePresentations: verifiablePresentationResults.map((r) =>
+        getW3cVerifiablePresentationInstance(r.verifiablePresentation)
+      ),
+      presentationSubmission,
+      presentationSubmissionLocation: firstVerifiablePresentationResult.presentationSubmissionLocation,
+    }
+  }
+
+  public getPresentationSignCallback(
+    agentContext: AgentContext,
+    verificationMethod: VerificationMethod,
+    includePresentationSubmissionInVp = true
+  ) {
+    const w3cCredentialService = agentContext.dependencyManager.resolve(W3cCredentialService)
+
+    return async (callBackParams: PresentationSignCallBackParams) => {
+      // The created partial proof and presentation, as well as original supplied options
+      const { presentation: presentationJson, options } = callBackParams
+      const { challenge, domain, nonce } = options.proofOptions ?? {}
+      const { verificationMethod: verificationMethodId } = options.signatureOptions ?? {}
+
+      let presentationToSignJson = presentationJson
+      if (!includePresentationSubmissionInVp) {
+        presentationToSignJson = {
+          ...presentationToSignJson,
+          presentation_submission: undefined,
+        }
+      }
+      const w3cPresentation = JsonTransformer.fromJSON(presentationToSignJson, W3cPresentation)
+
+      if (verificationMethodId && verificationMethodId !== verificationMethod.id) {
+        throw new AriesFrameworkError(
+          `Verification method from signing options ${verificationMethodId} does not match verification method ${verificationMethod.id}.`
+        )
+      }
+
+      // NOTE: we currently don't support mixed presentations, where some credentials
+      // are JWT and some are JSON-LD. It could be however that the presentation contains
+      // some JWT and some JSON-LD credentials. (for DDIP we only support JWT, so we should be fine)
+      const isJwt = typeof presentationJson.verifiableCredential?.[0] === 'string'
+
+      if (!isJwt) {
+        throw new AriesFrameworkError(`Only JWT credentials are supported for presentation exchange.`)
+      }
+
+      const key = getKeyFromVerificationMethod(verificationMethod)
+      const jwk = getJwkFromKey(key)
+
+      const alg = jwk.supportedSignatureAlgorithms[0]
+      if (!alg) {
+        throw new AriesFrameworkError(`No supported algs for key type: ${key.keyType}`)
+      }
+
+      const signedPresentation = await w3cCredentialService.signPresentation(agentContext, {
+        format: ClaimFormat.JwtVp,
+        verificationMethod: verificationMethod.id,
+        presentation: w3cPresentation,
+        alg,
+        challenge: challenge ?? nonce ?? utils.uuid(),
```

Comment:
we should not use a `uuid` as a challenge. `agenContext.wallet.generateNonce` can be used.